sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -266,6 +266,13 @@ trait CheckAnalysis extends PredicateHelper {
                 s"but one table has '${firstError.output.length}' columns and another table has " +
                 s"'${s.children.head.output.length}' columns")
 
+          case Union(c) if c.exists(_.output.map(_.dataType) != c.head.output.map(_.dataType)) =>
+            val firstError = c.find(_.output.map(_.dataType) != c.head.output.map(_.dataType)).get
+            failAnalysis(
```

Comment:
It would be very useful to know which field(s) are causing trouble. Now I have to scan the list my self.